src/webp/vp8.rs

```diff
@@ -884,33 +932,38 @@ impl<R: Read> VP8Decoder<R> {
         }
     }
 
-    fn init_partitions(&mut self, n: usize) -> io::Result<()> {
+    fn init_partitions(&mut self, n: usize) -> ImageResult<()> {
         if n > 1 {
-            let mut sizes = Vec::with_capacity(3 * n - 3);
-            try!(
-                self.r
-                    .by_ref()
-                    .take(3 * n as u64 - 3)
-                    .read_to_end(&mut sizes)
-            );
+            let mut sizes = vec![0; 3 * n - 3];
+            self.r.read_exact(sizes.as_mut_slice())?;
 
             for (i, s) in sizes.chunks(3).enumerate() {
-                let size = u32::from(s[0]) + ((u32::from(s[1])) << 8) + ((u32::from(s[2])) << 16);
-                let mut buf = Vec::with_capacity(size as usize);
-                try!(self.r.by_ref().take(u64::from(size)).read_to_end(&mut buf));
+                let size = {s}.read_u24::<LittleEndian>()
+                    .expect("Reading from &[u8] can't fail and the chunk is complete");
+
+                let mut buf = vec![0; size as usize];
+                self.r.read_exact(buf.as_mut_slice())?;
 
-                self.partitions[i].init(buf);
+                self.partitions[i].init(buf)?;
             }
         }
 
         let mut buf = Vec::new();
-        try!(self.r.read_to_end(&mut buf));
-        self.partitions[n - 1].init(buf);
+        self.r.read_to_end(&mut buf)?;
```

Comment:
This looks sort of weird. Shouldn't the decoder know how large the last partition is, since more frames can follow? I would guess this easily could get some sort of memory exhaustion error?

Reply:
No, the last partition is actually implicitely sized by the length of the upper partition in the `RIFF` file. In this particular instance, the `Read` instance should be instantiated with a `file.take(riff.len)` in the `webp` decoder. The reference implementation keeps track of all remaining bytes by propagating a `unsigned int sz` into every function call instead, which is effectively the same.

Reply:
Hm, it seems that this is however not the case in the upper layer: https://github.com/image-rs/image/blob/445b60bcd6e5578f28d1c81fffa37187877a2554/src/webp/decoder.rs#L70

Reply:
Will fix that, and include it as a test case.

Reply:
But is still don't completely understand. I thought `read_frame_header` would run once for each of the possibly very many frames in a vp8 stream, and then `read_to_end` could be called many times?

Reply:
From the reference:
> At the highest level, every compressed frame has three or more
   pieces.  It begins with an uncompressed data chunk comprising
   10 bytes in the case of key frames and 3 bytes for interframes.  This
   is followed by two or more blocks of compressed data (called
   partitions).  These compressed data partitions begin and end on byte
   boundaries.

To me this seems like the partitions only last until the next frame header?

Reply:
From [section-9.5](https://tools.ietf.org/html/rfc6386#section-9.5):
> Offsets are embedded in the bitstream to provide the decoder direct
   access to token partitions.  If the number of data partitions is
   greater than 1, the size of each partition (except the last) is
   written in 3 bytes (24 bits).  The size of the last partition is the
   remainder of the data not used by any of the previous partitions.

I was under the impression that there was one `VP8` [chunk](https://developers.google.com/speed/webp/docs/riff_container#simple_file_format_lossy) in the `RIFF` for every 'compressed frame'. Otherwise, those two statements don't exactly make sense to me. For image (single-frame vp8) this doesn't make a difference, the only contained frame in the chunk is the image.

Reply:
But I didn't really get it completely, and am not really sure. However, it seems to work.

Reply:
Yeah, that makes sense. So it is up to the container format so define frame boundaries, and the vp8 decoder just decodes whatever data is given to it as a single frame?

Reply:
It seems so? At least then it makes sense that there is seemingly one length number too few in the frame header.